src/features/embalm/stepContent/hooks/useCreateSarcophagus/useUploadFileAndKeyShares.ts

```diff
@@ -0,0 +1,96 @@
+import { useCallback, useContext } from 'react';
+import { encrypt, readFileDataAsBase64 } from '../../../../../lib/utils/helpers';
+import useArweaveService from '../../../../../hooks/useArweaveService';
+import { useSelector } from '../../../../../store';
+import { CreateSarcophagusContext } from '../../context/CreateSarcophagusContext';
+import {
+  encryptShardsWithArchaeologistPublicKeys,
+  encryptShardsWithRecipientPublicKey,
+} from '../../utils/createSarcophagus';
+import { split } from 'shamirs-secret-sharing-ts';
+
+export function useUploadFileAndKeyShares() {
+  const { uploadToArweave } = useArweaveService();
+  const { file, recipientState } = useSelector(x => x.embalmState);
+  const { selectedArchaeologists, requiredArchaeologists } = useSelector(x => x.embalmState);
+  const { outerPrivateKey, outerPublicKey, archaeologistPublicKeys, setSarcophagusPayloadTxId } =
+    useContext(CreateSarcophagusContext);
+
+  const uploadAndSetArweavePayload = useCallback(async () => {
+    try {
+      const data = await readFileDataAsBase64(file!);
+      const payload = {
+        fileName: file?.name,
+        data,
+      };
+
+      /**
+       * File upload data
+       */
+      // Step 1: Encrypt the payload with the generated keypair
+      const encryptedOuterLayer = await encrypt(
+        outerPublicKey!,
+        Buffer.from(JSON.stringify(payload), 'hex')
+      );
+
+      /**
+       * Double encrypted keyshares upload data
+       */
+      // Step 1: Split the outer layer private key using shamirs secret sharing
+      const keyShares: Uint8Array[] = split(outerPrivateKey, {
+        shares: selectedArchaeologists.length,
+        threshold: requiredArchaeologists,
+      });
+
+      // Step 2: Encrypt each shard with the recipient public key
+      const keySharesEncryptedInner = await encryptShardsWithRecipientPublicKey(
+        recipientState.publicKey,
+        keyShares
+      );
+
+      // Step 3: Encrypt each shard again with the arch public keys
+      const keySharesEncryptedOuter = await encryptShardsWithArchaeologistPublicKeys(
+        Array.from(archaeologistPublicKeys.values()),
+        keySharesEncryptedInner
+      );
+
+      /**
+       * Format data for upload
+       */
+      const doubleEncryptedKeyShares: Record<string, string> = keySharesEncryptedOuter.reduce(
+        (acc, keyShare) => ({
+          ...acc,
+          [keyShare.publicKey]: keyShare.encryptedShard,
+        }),
+        {}
+      );
+
+      // TODO: #multiple-key-update - this affects resurrection, may need to change (and be tested)
+      const combinedPayload = {
+        file: encryptedOuterLayer,
+        keyShares: Buffer.from(JSON.stringify(doubleEncryptedKeyShares)),
```

Comment:
Since 75 stringifies and `Buffer.from`s the `combinedPayload` data anyway, is this conversion needed? Might be possible to have one less parsing step to deal with during resurrection

Reply:
This was the code I was least sure about -- I only did this b/c the `encryptedOuterLayer` is a buffer so wanted to match that format. But once we can actually test this, will probably need cleaned up.

Reply:
@sethhrbek I tested this out and the extra conversion isn't needed. Although the resurrection function had to similarly account for it. IOW removing this extra step also removes the other-way-round conversion step during decryption.

I've pushed an update.

Reply:
@DarksightKellar -- awesome, thanks for looking into that and making the updates.